File: zh-hant/codes/csharp/chapter_tree/binary_tree.cs
/**
 * File: binary_tree.cs
 * Created Time: 2022-12-23
 * Author: haptear (haptear@hotmail.com)
 */

namespace hello_algo.chapter_tree;

public class binary_tree {
    [Test]
    public void Test() {
        /* 初始化二元樹 */
        // 初始化節點
        TreeNode n1 = new(1);
        TreeNode n2 = new(2);
        TreeNode n3 = new(3);
        TreeNode n4 = new(4);
        TreeNode n5 = new(5);
        // 構建節點之間的引用（指標）
        n1.left = n2;
        n1.right = n3;
        n2.left = n4;
        n2.right = n5;
        Console.WriteLine("\n初始化二元樹\n");
        PrintUtil.PrintTree(n1);

        /* 插入與刪除節點 */
        TreeNode P = new(0);
        // 在 n1 -> n2 中間插入節點 P
        n1.left = P;
        P.left = n2;
        Console.WriteLine("\n插入節點 P 後\n");
        PrintUtil.PrintTree(n1);
        // 刪除節點 P
        n1.left = n2;
        Console.WriteLine("\n刪除節點 P 後\n");
        PrintUtil.PrintTree(n1);
    }
}

File: vi/codes/rust/chapter_tree/binary_tree.rs
/**
 * File: binary_tree.rs
 * Created Time: 2023-02-27
 * Author: xBLACKICEx (xBLACKICE@outlook.com)
 */
use std::rc::Rc;
use hello_algo_rust::include::{print_util, TreeNode};

/* Driver Code */
fn main() {
    /* Initialize binary tree */
    // Initialize nodes
    let n1 = TreeNode::new(1);
    let n2 = TreeNode::new(2);
    let n3 = TreeNode::new(3);
    let n4 = TreeNode::new(4);
    let n5 = TreeNode::new(5);
    // Build references (pointers) between nodes
    n1.borrow_mut().left = Some(Rc::clone(&n2));
    n1.borrow_mut().right = Some(Rc::clone(&n3));
    n2.borrow_mut().left = Some(Rc::clone(&n4));
    n2.borrow_mut().right = Some(Rc::clone(&n5));
    println!("\nInitialize binary tree\n");
    print_util::print_tree(&n1);

    // Insert node and delete node
    let p = TreeNode::new(0);
    // Delete node
    p.borrow_mut().left = Some(Rc::clone(&n2));
    n1.borrow_mut().left = Some(Rc::clone(&p));
    println!("\nAfter inserting node P\n");
    print_util::print_tree(&n1);
    // Remove node P
    drop(p);
    n1.borrow_mut().left = Some(Rc::clone(&n2));
    println!("\nAfter removing node P\n");
    print_util::print_tree(&n1);
}
